source/fundamentals/typescript.txt

```diff
@@ -0,0 +1,97 @@
+==========
+TypeScript
+==========
+
+.. default-domain:: mongodb
+
+.. contents:: On this page
+   :local:
+   :backlinks: none
+   :depth: 2
+   :class: singlecols
+
+Overview
+--------
+
+In this guide, you can learn about the **TypeScript** features and limitiations
+of the MongoDB Node.js driver. TypeScript is a strongly typed programming
+language that compiles to JavaScript.
+
+All TypeScript features of the driver are optional. All valid JavaScript
+code written with the driver is also valid TypeScript code. 
+
+For more information, see the
+`TypeScript website <https://www.typescriptlang.org/>`__.
+
+Features
+--------
+
+If you use TypeScript, you can specify a type for some classes in the driver.
+All classes that accept a type paramater in the driver have the default type
+:node-api-4.0:`Document <interfaces/document.html>`.
+
+The following classes accept any type that extends the ``Document``
+interface:
+
+- :node-api-4.0:`Collection <classes/collection.html>`
+- :node-api-4.0:`ChangeStream <classes/changestream.html>`
+
+You can pass a type parameter that extends the ``Document`` interface like this:
+
+.. literalinclude:: /code-snippets/typescript/extend-document.ts
+   :language: typescript
+   :linenos:
+
+The following classes accept any type parameter:
+
+- :node-api-4.0:`FindCursor <classes/findcursor.html>`
+- :node-api-4.0:`AggregationCursor <classes/aggregationcursor.html>`
+
+You can find a code snippet that shows how to specify a type for the ``FindCursor``
+class in the
+:ref:`Find a Document Usage Example <node-driver-findone-usage-example-code-snippet>`.
+
+Limitations
+-----------
+
+.. _node-driver-typescript-limitations-dot-notation:
+
+The driver cannot infer the type of values with keys containing **dot
+notation**. Dot notation is a property access syntax for navigating BSON objects.
+Click on the tabs to see code snippets that highlight this behavior:
+
+.. tabs::
+
+  .. tab:: Dot Notation
+     :tabid: dot-notation
+  
+     The following code snippet does not raise a type error:
+
+     .. literalinclude:: /code-snippets/typescript/dot-notation.ts
+        :language: typescript
+        :linenos:
+        :start-after: start-no-error
+        :end-before: end-no-error
+
+  .. tab:: Nested Objects
+     :tabid: nested-objects
+    
+     The following code snippet raises a type error:
+
+     .. literalinclude:: /code-snippets/typescript/dot-notation.ts
+        :language: typescript
+        :linenos:
+        :start-after: start-error
+        :end-before: end-error
+
+     This is the error:
+
+     .. code-block:: text
+
+        Type 'string' is not assignable to type 'number'.
+
+Despite the lack of type safety, you should still use dot notation to access nested
```

Comment:
```suggestion
Despite the lack of type safety, we still recommend using dot notation to access nested
```

Reply:
Sounds good! One thing to be aware of for saying "we recommend" in docs is that this becomes an opinion on behalf of MongoDB the company. Will make the edit, just giving context around the specific meaning of "we recommend".

[Passage from style guide](https://docs.mongodb.com/meta/style-guide/writing/write-to-the-user/#id1):

>  Use we recommend rather than it’s recommended or MongoDB recommends. Limit how often you state recommendations. This expresses an opinion on behalf of MongoDB. Make sure that the engineering team or TSE team has expressed this opinion before stating anything as a recommendation. You should be prepared to defend any recommendations you publish with a ticket number from engineering or the TSE organization.

Reply:
Saying you "should use" seems like a recommendation without saying the word so we need to make this defensible regardless.

Maybe we can just add why we recommend it here. Using dot notation vs nested documents leads to smaller BSON payloads over the network. There may be more reasons but at least this one is a guarantee. There's no ticket number for that but its inherent to the BSON format.

Typescript folks will likely want to spell out their nested documents for the sake of type safety so I also don't want to scare people off from that because while there is overhead it may be negligible and worth the downside to maintain developer productivity.

Reply:
Added a note on why we recommend it! Lead to a large internal DevEd team thread, and it is fairly difficult to mimic the behavior or dot notation using query expressions. You have  the ENTIRE subdocument in a query must match, making not possible to query on a single field reliably (you can't return documents with more or less fields). Nathan Leniz came up with a query that behaves like dot notation and it is hideous (but impressive that he came up with it all!) that I think demonstrates why dot notation is preferable. 

Query:

 collection.find({"$jsonSchema": { required: [ "field" ], properties: { field: { bsonType: "object", properties: { s1: { enum: ["hi"] }}}}}})

Reply:
That's quite the query! I did forget that the whole subdocument must match so yea dot notation is certainly a must. changes LGTM